evm-typegen/src/main.ts

```diff
@@ -0,0 +1,153 @@
+    output.line(`import inputJson from "${path.resolve(inputPathRaw)}";`);
+    output.line("");
+    output.line("const abi = new Interface(inputJson);");
+    output.line("");
+
+    const eventTypeIndexes: { [key: string]: number } = {};
+
+    const abiEvents = rawABI.filter((decl: any): boolean => decl.type === "event").map((decl: any): any => {
+        let signature = `${decl.name}(`;
+    
+        if(decl.inputs.length > 0) {
+            signature += decl.inputs[0].type;
+        }
+    
+        for (let i=1; i<decl.inputs.length; ++i) {
+            const input = decl.inputs[i];
+            signature += `,${input.type}`;
+        }
+    
+        signature += ")";
+    
+        decl.signature = signature;
+        if (eventTypeIndexes[decl.name] === undefined) {
+            eventTypeIndexes[decl.name] = 0;
+        } else {
+            eventTypeIndexes[decl.name]++;
+        }
+        decl.eventTypeName = decl.name + eventTypeIndexes[decl.name].toString();
+    
+        return decl;
+    });
+
+    const abiCoder = new AbiCoder();
+
+    for(const decl of abiEvents) {
+        output.block(`export interface ${decl.eventTypeName}Event`, () => {
+            for (const input of decl.inputs) {
+                output.line(`\t${input.name}: ${getType(input)};`);
+            }
+        });
+        output.line("");
+    }
+
+    output.block("export const events =", () => {
+        for(const decl of abiEvents) {
+            output.block(`"${decl.signature}": `, () => {
+                output.line(`topic: abi.getEventTopic("${decl.signature}"),`);
+                output.block(`decode(data: EvmLogHandlerContext): ${decl.eventTypeName}Event`, () => {
+                    output.line(`const result = abi.decodeEventLog(`);
+                    output.line(`\tabi.getEvent("${decl.signature}"),`);
+                    output.line(`\tdata.data || "",`);
+                    output.line("\tdata.topics");
+                    output.line(");");
+                    output.block("return ", () => {
+                        for (let i=0; i<decl.inputs.length; ++i) {
+                            const input = decl.inputs[i];
+                            output.line(`${input.name}: ${`result[${i}]`},`);
+                        }
+                    });
+                });
+            });
+            output.line(",");
+        }
+    });
+
+    fs.writeFileSync(outputPathRaw, output.toString());
+}
+
+// taken from: https://github.com/ethers-io/ethers.js/blob/948f77050dae884fe88932fd88af75560aac9d78/packages/cli/src.ts/typescript.ts#L10
+function getType(param: any, flexible?: boolean): string {
+    if (param.type === "address" || param.type === "string") { return "string"; }
+
+    if (param.type === "bool") { return "boolean" }
+
+    if (param.type.substring(0, 5) === "bytes") {
+        if (flexible) {
+            return "string | ethers.utils.BytesLike";
```

Comment:
`ethers` library is really huge, on the other hand, I think we should convert everything to native js types.

Reply:
broke down the depedency to two smaller libraries